gcs/constraints/smart_table.cc

```diff
@@ -0,0 +1,642 @@
+                const auto unrestricted = get_unrestricted(vars, tuples[tuple_idx]);
+                for (const auto & var : unrestricted) {
+                    unsupported[var] = vector<Integer>{};
+                }
+            }
+        }
+
+        bool some_tuple_still_feasible = false;
+        for (unsigned int tuple_idx = 0; tuple_idx < tuples.size(); ++tuple_idx) {
+            if (state.optional_single_value(selectors[tuple_idx]) != 0_i) {
+                some_tuple_still_feasible = true;
+            }
+        }
+        if (! some_tuple_still_feasible) {
+            contradiction = true;
+        }
+
+        for (const auto & var : vars) {
+            for (const auto & value : unsupported[var]) {
+                switch (state.infer_not_equal(var, value, JustifyUsingRUP{})) {
+                case Inference::NoChange: break;
+                case Inference::Change: changed = true; break;
+                case Inference::Contradiction: contradiction = true; break;
+                }
+            }
+        }
+
+        return contradiction
+            ? Inference::Contradiction
+            : changed ? Inference::Change
+                      : Inference::NoChange;
+    }
+
+    auto build_tree(const IntegerVariableID & root, int current_level, vector<vector<SmartEntry>> & entry_tree,
+        unordered_map<IntegerVariableID, bool> & node_visited,
+        const unordered_map<IntegerVariableID, vector<SmartEntry>> & adjacent_edges) -> void
+    {
+        node_visited[root] = true;
+
+        // Simple recursive traverse
+        // Note: Perhaps we should build the tree in a "smarter" form e.g. make sure var_1 is always the node
+        //       closer to the root.
+        for (const auto & edge : adjacent_edges.at(root)) {
+            overloaded{
+                [&](BinaryEntry binary_entry) {
+                    if (! node_visited[binary_entry.var_1]) {
+                        entry_tree[current_level].emplace_back(edge);
+                        entry_tree.emplace_back();
+
+                        build_tree(binary_entry.var_1, current_level + 1, entry_tree, node_visited, adjacent_edges);
+                    }
+                    else if (! node_visited[binary_entry.var_2]) {
+                        entry_tree[current_level].emplace_back(edge);
+                        entry_tree.emplace_back();
+                        build_tree(binary_entry.var_2, current_level + 1, entry_tree, node_visited, adjacent_edges);
+                    }
+                },
+                [&](const UnarySetEntry &) {
+                    entry_tree[current_level].emplace_back(edge);
+                },
+                [&](const UnaryValueEntry &) {
+                    entry_tree[current_level].emplace_back(edge);
+                }}
+                .visit(edge);
+        }
+    }
+
+    auto build_forests(SmartTuples & tuples) -> vector<SmartTable::Forest>
```

Comment:
```suggestion
    [[nodiscard]] auto build_forests(SmartTuples & tuples) -> vector<SmartTable::Forest>
```

Reply:
Fixed in c62e8914eb9e01df2a147f01713ff505d7e905e0